contrib/opencensus-ext-azure/opencensus/ext/azure/metrics_exporter/__init__.py

```diff
@@ -12,30 +12,94 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import threading
-
 from opencensus.ext.azure.common import Options
+from opencensus.metrics.export.metric_descriptor import MetricDescriptorType
 from opencensus.metrics import transport
 from opencensus.stats import stats
+from opencensus.ext.azure.common import utils
+from opencensus.ext.azure.common.protocol import Data
+from opencensus.ext.azure.common.protocol import DataPoint
+from opencensus.ext.azure.common.protocol import Envelope
+from opencensus.ext.azure.common.protocol import MetricData
+from opencensus.ext.azure.common.transport import TransportMixin
 
 __all__ = ['MetricsExporter', 'new_metrics_exporter']
 
 
-class MetricsExporter(object):
+class MetricsExporter(TransportMixin):
     """Metrics exporter for Microsoft Azure Monitor."""
 
-    def __init__(self, options):
+    def __init__(self, options=None):
+        if options is None:
+            options = Options()
         self.options = options
         if not self.options.instrumentation_key:
             raise ValueError('The instrumentation_key is not provided.')
-        self._md_cache = {}
-        self._md_lock = threading.Lock()
+        # TODO: Implement retry logic
+        self.storage = None
 
     def export_metrics(self, metrics):
-        metrics = list(metrics)
-        for metric in metrics:
-            # TODO: implement export to Azure Monitor functionality
-            print(repr(metric))
+        if metrics:
+            for metric in metrics:
+                # No support for histogram aggregations
+                type_ = metric.descriptor.type
+                if type_ != MetricDescriptorType.CUMULATIVE_DISTRIBUTION:
+                    md = metric.descriptor
+                    # Each time series will be uniquely
+                    # identified by it's label values
+                    for time_series in metric.time_series:
+                        # Using stats, time_series should
+                        # only have one point which contains
+                        # the aggregated value
+                        data_point = self.create_data_points(
+                                     time_series, md)[0]
+                        # The timestamp is when the metric was recorded
+                        time_stamp = time_series.points[0].timestamp
+                        # Get the properties using label keys from metric
+                        # and label values of the time series
+                        properties = self.create_properties(time_series, md)
+                        envelope = self.create_envelope(
+                                   data_point, time_stamp, properties)
+                        self._transmit(envelope)
```

Comment:
This wouldn't work.
`_transmit` takes a list of envelopes, passing a single envelope would always fail with illegal request body.
This also shows how important it is to check the result instead of silently ignore the error.

Reply:
I tested this and it works fine. _transmit merely calls json.dumps and the ingestion service is able to accept individual envelopes. I am able to receive a 200 status whether I input envelope or a list of envelopes. However, given that the name is "envelopes" with a plural, I can change it to pass in a list. Also this me be appropriate if I was mistaken about the ingestion service taking only one data point at a time (I will look into this).